### ISSUE TITLE
LD_LIBRARY_PATH issues resulting in Wine client error:0: version mismatch errors

### ISSUE THREAD

**Issue by 00Asgaroth00** [lutris/lutris#903]
I'm having this issue with lutris v0.4.18, installed using Arch AUR package on Manjaro linux (Arch variant).

My system wine installed is wine-staging-3.8. If I use anything newer than that I end up with client version mismatch.

For example, using the [Blizzard Battle.net app](https://lutris.net/games/battlenet/) installer from Lutris's web page, the installer installs wine version pba-firerat-3.9-x86_64. When it attempts to start the installation it fails with:
```
DEBUG    2018-06-05 13:52:17,381 [thread]:Running command: "/home/username/.local/share/lutris/runners/wine/pba-firerat-3.9-x86_64/bin/wine" "/home/username/.cache/lutris/installer/battlenet/Battle.net-Setup.exe"
DEBUG    2018-06-05 13:52:17,381 [thread]:Storing environment variable WINEARCH to win64
DEBUG    2018-06-05 13:52:17,382 [thread]:Storing environment variable WINE to /home/username/.local/share/lutris/runners/wine/pba-firerat-3.9-x86_64/bin/wine
DEBUG    2018-06-05 13:52:17,382 [thread]:Storing environment variable WINEPREFIX to /data/lutris/battlenet
DEBUG    2018-06-05 13:52:17,382 [thread]:Storing environment variable LD_LIBRARY_PATH to /usr/lib:/usr/lib32:/home/username/.local/share/lutris/runtime/lib32:/home/username/.local/share/lutris/runtime/steam/i386/lib/i386-linux-gnu:/home/username/.local/share/lutris/runtime/steam/i386/lib:/home/username/.local/share/lutris/runtime/steam/i386/usr/lib/i386-linux-gnu:/home/username/.local/share/lutris/runtime/steam/i386/usr/lib:/home/username/.local/share/lutris/runtime/lib64:/home/username/.local/share/lutris/runtime/steam/amd64/lib/x86_64-linux-gnu:/home/username/.local/share/lutris/runtime/steam/amd64/lib:/home/username/.local/share/lutris/runtime/steam/amd64/usr/lib/x86_64-linux-gnu:/home/username/.local/share/lutris/runtime/steam/amd64/usr/lib
wine client error:0: version mismatch 554/549.
```
The wine command above is attempting to pull libwine from the system's installed wine, as demonstrated below:
```
~]$ LD_LIBRARY_PATH="/usr/lib:/usr/lib32:/home/username/.local/share/lutris/runtime/lib32:/home/username/.local/share/lutris/runtime/steam/i386/lib/i386-linux-gnu:/home/username/.local/share/lutris/runtime/steam/i386/lib:/home/username/.local/share/lutris/runtime/steam/i386/usr/lib/i386-linux-gnu:/home/username/.local/share/lutris/runtime/steam/i386/usr/lib:/home/username/.local/share/lutris/runtime/lib64:/home/username/.local/share/lutris/runtime/steam/amd64/lib/x86_64-linux-gnu:/home/username/.local/share/lutris/runtime/steam/amd64/lib:/home/username/.local/share/lutris/runtime/steam/amd64/usr/lib/x86_64-linux-gnu:/home/username/.local/share/lutris/runtime/steam/amd64/usr/lib" ldd -r /home/username/.local/share/lutris/runners/wine/pba-firerat-3.9-x86_64/bin/wine
	linux-gate.so.1 (0xf7f28000)
	libwine.so.1 => /usr/lib32/libwine.so.1 (0xf7d58000)
	libpthread.so.0 => /usr/lib32/libpthread.so.0 (0xf7d39000)
	libc.so.6 => /usr/lib32/libc.so.6 (0xf7b5e000)
	libdl.so.2 => /usr/lib32/libdl.so.2 (0xf7b59000)
	/lib/ld-linux.so.2 => /usr/lib/ld-linux.so.2 (0xf7f2a000)
```
If I remove the prepended "/usr/lib" and "/usr/lib32" it then pulls the correct libwine as demonstrated below:
```
~]$ LD_LIBRARY_PATH="/home/username/.local/share/lutris/runtime/lib32:/home/username/.local/share/lutris/runtime/steam/i386/lib/i386-linux-gnu:/home/username/.local/share/lutris/runtime/steam/i386/lib:/home/username/.local/share/lutris/runtime/steam/i386/usr/lib/i386-linux-gnu:/home/username/.local/share/lutris/runtime/steam/i386/usr/lib:/home/username/.local/share/lutris/runtime/lib64:/home/username/.local/share/lutris/runtime/steam/amd64/lib/x86_64-linux-gnu:/home/username/.local/share/lutris/runtime/steam/amd64/lib:/home/username/.local/share/lutris/runtime/steam/amd64/usr/lib/x86_64-linux-gnu:/home/username/.local/share/lutris/runtime/steam/amd64/usr/lib" ldd -r /home/username/.local/share/lutris/runners/wine/pba-firerat-3.9-x86_64/bin/wine
	linux-gate.so.1 (0xf7f32000)
	libwine.so.1 => /home/username/.local/share/lutris/runners/wine/pba-firerat-3.9-x86_64/bin/../lib/libwine.so.1 (0xf7d61000)
	libpthread.so.0 => /usr/lib32/libpthread.so.0 (0xf7d08000)
	libc.so.6 => /usr/lib32/libc.so.6 (0xf7b2d000)
	libdl.so.2 => /usr/lib32/libdl.so.2 (0xf7b28000)
	/lib/ld-linux.so.2 => /usr/lib/ld-linux.so.2 (0xf7f34000)
```
I had a look at the runtime.py that is mentioned in a similar issue (#875) and it appears that /usr/lib and /usr/lib32 are manually prepended to the library path, as seen below:

```
   149          # Put /usr/lib at the beginning, this prioritizes system libraries over
   150          # the Lutris and Steam runtimes.
   151          paths = ["/usr/lib"]
   152          if os.path.exists("/usr/lib32"):
   153              # Also let the system take over 32bit libs.
   154              paths.append("/usr/lib32")
   155  
   156          # Then resolve absolute paths for the runtime
   157          paths += [os.path.join(RUNTIME_DIR, path) for path in runtime_paths]
   158  
   159      if os.environ.get('LD_LIBRARY_PATH'):
   160          paths.append(os.environ['LD_LIBRARY_PATH'])
```
I manually updated runtime.py to append these paths to the library path as below:
```
   149          # Put /usr/lib at the beginning, this prioritizes system libraries over
   150          # the Lutris and Steam runtimes.
   151          sys_paths = ["/usr/lib"]
   152          if os.path.exists("/usr/lib32"):
   153              # Also let the system take over 32bit libs.
   154              sys_paths.append("/usr/lib32")
   155  
   156          # Then resolve absolute paths for the runtime
   157          paths += [os.path.join(RUNTIME_DIR, path) for path in runtime_paths]
   158          paths += [p for p in sys_paths]
   159  
   160      if os.environ.get('LD_LIBRARY_PATH'):
   161          paths.append(os.environ['LD_LIBRARY_PATH'])
```
This, however, does *not* fix the issue, it still pulls the libwine library from /usr/lib32.

If I do not append the /usr/lib|/usr/lib32 paths at all, then the installer works, and I do not get the version mismatch, as below:
```
   149          # Put /usr/lib at the beginning, this prioritizes system libraries over
   150          # the Lutris and Steam runtimes.
   151          sys_paths = ["/usr/lib"]
   152          if os.path.exists("/usr/lib32"):
   153              # Also let the system take over 32bit libs.
   154              sys_paths.append("/usr/lib32")
   155  
   156          # Then resolve absolute paths for the runtime
   157          paths += [os.path.join(RUNTIME_DIR, path) for path in runtime_paths]
   158          # paths += [p for p in sys_paths]
   159  
   160      if os.environ.get('LD_LIBRARY_PATH'):
   161          paths.append(os.environ['LD_LIBRARY_PATH'])
```
I guess this works as it cannot find it in the global library path and defaults to looking in the "current directory" that the application is being called from.

I wonder, is it possible, from lutris, to find out what lutris wine version is being used and then, manually prepending the lutris wine path to the ld_library_path?

**Comment by nonchip** [comment]
when i mentioned you might want to open a new issue because it doesn't have much to do with the wine version mismatch i also meant you might call it something like "LD_LIBRARY_PATH generation broken" not literally the title of the issue it wasn't related to :P

**Comment by 00Asgaroth00** [comment]
OK, title updated.

I thought this might be a give-away:
```
libwine.so.1 => /home/username/.local/share/lutris/runners/wine/pba-firerat-3.9-x86_64/bin/../lib/libwine.so.1 (0xf7d61000)
```
Where it is using the currently directory for the app, and issuing a "../lib" as a path for 32-bit libraries. Is that a compile time option then (-r path) where it searches if it cannot find it locally? 

I dont see the path for this wine version being set automatically here:
```
DEBUG    2018-06-05 13:52:17,382 [thread]:Storing environment variable LD_LIBRARY_PATH to /usr/lib:/usr/lib32:/home/username/.local/share/lutris/runtime/lib32:/home/username/.local/share/lutris/runtime/steam/i386/lib/i386-linux-gnu:/home/username/.local/share/lutris/runtime/steam/i386/lib:/home/username/.local/share/lutris/runtime/steam/i386/usr/lib/i386-linux-gnu:/home/username/.local/share/lutris/runtime/steam/i386/usr/lib:/home/username/.local/share/lutris/runtime/lib64:/home/username/.local/share/lutris/runtime/steam/amd64/lib/x86_64-linux-gnu:/home/username/.local/share/lutris/runtime/steam/amd64/lib:/home/username/.local/share/lutris/runtime/steam/amd64/usr/lib/x86_64-linux-gnu:/home/username/.local/share/lutris/runtime/steam/amd64/usr/lib
```
The path for the lutris installed wine version for this prefix is this:
```
/home/username/.local/share/lutris/runners/wine/pba-firerat-3.9-x86_64
```
Which I dont see being set in the LD_LIBRARY_PATH at all, not sure where it is being set automatically.

What I am asking is if we can see the installation path of the wine version being used in lutris and then prepending, for example,  <wine_path>/lib|lib64 to the LD_LIBRARY_PATH.

**Comment by 00Asgaroth00** [comment]
Just another comment on the get_paths() function:

```
def get_paths():
    """Return a list of paths containing the runtime libraries."""
    paths = []

    if not RUNTIME_DISABLED:
        runtime_paths = [
            "lib32",
            "steam/i386/lib/i386-linux-gnu",
            "steam/i386/lib",
            "steam/i386/usr/lib/i386-linux-gnu",
            "steam/i386/usr/lib"
        ]

        if system.IS_64BIT:
            runtime_paths += [
                "lib64",
                "steam/amd64/lib/x86_64-linux-gnu",
                "steam/amd64/lib",
                "steam/amd64/usr/lib/x86_64-linux-gnu",
                "steam/amd64/usr/lib"
            ]

        # Put /usr/lib at the beginning, this prioritizes system libraries over
        # the Lutris and Steam runtimes.
        paths = ["/usr/lib"]
        if os.path.exists("/usr/lib32"):
            # Also let the system take over 32bit libs.
            paths.append("/usr/lib32")

        # Then resolve absolute paths for the runtime
        paths += [os.path.join(RUNTIME_DIR, path) for path in runtime_paths]

    if os.environ.get('LD_LIBRARY_PATH'):
        paths.append(os.environ['LD_LIBRARY_PATH'])

    return paths
```
the variable "paths" is initially being defined as a list
```
paths = []
```
It is then used by the system paths to initialize/overwrite and then append to this list:
```
paths = ["/usr/lib"]
        if os.path.exists("/usr/lib32"):
            # Also let the system take over 32bit libs.
            paths.append("/usr/lib32")
```
then is being converted to a string to be returned:
```
paths += [os.path.join(RUNTIME_DIR, path) for path in runtime_paths]
```
Then referenced in list format again to append any pre-existing library paths:
```
if os.environ.get('LD_LIBRARY_PATH'):
        paths.append(os.environ['LD_LIBRARY_PATH'])
```
I'm surprised this has not caused issues, or, does the python interpreter magically create two variables behind the scenes?

**Comment by 00Asgaroth00** [comment]
Would it be possible to add function arguments for get_env() and get_paths() that would specify the base directory and architecture of the runner being used, then, we could properly prepend the library directories when generating the paths.

That way you could prepend the runner library paths and still append the /usr/lib and /usr/lib32 library paths if they are required for other system libraries?

Looks like get_env() is called once in thread.py, once in command.py and once in runner.py.

Looks like get_paths is called once in runtime.py (by get_env()) and once in wine.py.

I'm no python expert but it doesnt look like that may be an intrusive change, however, I'm not sure if the wine runner base directory is available in any of those locations to pass over to the runtime.py functions.

**Comment by nonchip** [comment]
> then, we could properly prepend the library directories when generating the paths.

no we couldn't. when using 32bit components in 64bit wineprefixes (e.g. most of steam's internals are still WoW64) their dependencies (e.g. libpulse for audio, libgl for graphics etc) still need to be 32bit.

**Comment by 00Asgaroth00** [comment]
> no we couldn't. when using 32bit components in 64bit wineprefixes (e.g. most of steam's internals are still WoW64) their dependencies (e.g. libpulse for audio, libgl for graphics etc) still need to be 32bit.

How do we fix this then if we dont/cant set the proper library paths? If we want to pull the correct libraries for the runner being used then we need to setup the correct paths.

As it stands, if you system wine is older than your runner and your system libraries are prepended before the runner path then it will cause this issue as demonstrated in the initial post.

There is nothing stopping us from setting up both 32/64-bit paths when we have the base directory of the runner, at least, in this case you are ensuring that the runner is pulling it's own libraries that it was built with and not the system ones.

My machine just recieved an update (last night) for wine-staging 3.9, but I can almost guarentee it will pull in my system's libwine instead of the firerat libwine, the client mismatch error will be gone, but if a runner is built with custom patches, this may cause issues.

**Comment by nonchip** [comment]
> How do we fix this then if we dont/cant set the proper library paths? If we want to pull the correct libraries for the runner being used then we need to setup the correct paths.

by setting the paths in the correct order and/or places instead of omitting them selectively.

> As it stands, if you system wine is older than your runner and your system libraries are prepended before the runner path then it will cause this issue as demonstrated in the initial post.

same if they're newer btw. `/usr/lib[32]` shouldn't be prepended as-is (and actually probably shouldn't even have to be included, since `/etc/ld.so.conf` is a thing).

> There is nothing stopping us from setting up both 32/64-bit paths when we have the base directory of the runner, at least, in this case you are ensuring that the runner is pulling it's own libraries that it was built with and not the system ones.

actually as far as i know that should already be a thing. i see only runtime paths in your log above, not the runner ones. that's definitely something that {s,c}hould be fixed. this might help with wine specific issues, wouldn't fix the overall problem of the system libs in the environment messing with the load order though.

i think actually the reason to prepend it was the fact some libs in the runtime break when there are better/newer/whatever versions of them in the system paths. probably would be better to fix those libs or figure out a way to configure which libs should be pulled from the runtime (e.g. by having lutris manage a folder per game with symlinks to the runtime libs for the enabled ones; could still default to "all" but then allow you to disable them individually, or even detect lib errors like unresolved symbols etc in the stderr log and do it automatically) than abusing the environment variable like it's currently done.

**Comment by 00Asgaroth00** [comment]
> actually as far as i know that should already be a thing. i see only runtime paths in your log above, not the runner ones. that's definitely a bug that {s,c}hould be fixed. this might help with wine specific issues, wouldn't fix the overall problem of the system libs in the environment messing with the load order though.

That is exactly my point, they are the steam runtime library paths (in get_paths function in runtime.py), they do **not** include the wine build library paths. They **should** include the lutris wine build library paths.

I think I am getting my lutris terminology mixed up, I have been saying runner, when I actually mean the lutris wine build. I guess that could be considered a runner too, right?

The logs above are taken from an attempt to run an installation script from lutris.net, but, I guess the same applies to wine prefixes being started too (The same thing happened to me when attempting to start a staging 3.9 battlefield 1 prefix)



> add the runner paths (before the runtime ones)

Are these steam runtime paths even needed for non-steam runners? Maybe those should only be applied when the runner is winesteam?

**Comment by 00Asgaroth00** [comment]
okay, I guess at this point we probably need a lutris developer to pop into the discussion to see what they make of all of the above :)

**Comment by strycore** [comment]
I see what is happening, that's an oversight from a recent change in how the runtime is handled.

In the last releases, we started prioritizing /usr/lib and /usr/lib32 over the Lutris + Steam runtime because that solves most of the runtime issues.

The problem is that when Wine is installed system wide, the libwine.so in /usr/lib takes precedence over the one that should be used.

I would like to keep this prioritization going on so I suggest adding wine's lib folder in LD_LIBRARY_PATH at runtime, which should prevent this issue from happening.

**Comment by 00Asgaroth00** [comment]
Thanks, just for completeness, when **pre**pending the lutris wine build path, we should probably include the "wine" directory too, for example:

32-bit
```
<lutris wine build path>/lib
<lutris wine build path>/lib/wine
```
64-bit
```
<lutris wine build path>/lib64
<lutris wine build path>/lib64/wine
```

**Comment by nonchip** [comment]
> I would like to keep this prioritization going on so I suggest adding wine's lib folder in LD_LIBRARY_PATH at runtime, which should prevent this issue from happening.

yes, prepending wine even before `/usr/*` should fix this issue, but on the other hand the whole "prepending the system to the runtime" resolves most issues but might also introduce others we might have to keep an eye out for. might have to go for a customizable runtime instead at some point. but yeah currently that's the best/easiest fix.

> Thanks, just for completeness, when prepending the lutris wine build path, we should probably include the "wine" directory too, for example:

I'm not sure those `.dll.so`s are handled by `ld.so` but it won't hurt to include the folder either, yeah.

though actually i wouldn't even check for the bittiness there, just include all 4 paths, since if a path doesn't exist (which would only be the case for `lib64` on a 32bit build) `ld.so` just ignores it.

**Comment by strycore** [comment]
I've pushed a fix to `next`, there's also an option to disable the system libs from being added.

**Comment by nonchip** [comment]
@strycore "prefer system libs" in that patch does influence wine libs though, which doesn't make much sense. when as the wine version isn't the system's you should always prepend its paths before whatever comes next (system libs, runtime libs or nothing; since the value of the user's selected LD_LIBRARY_PATH (from the env table in sys settings) should be always before any of the autogenerated stuff).

**Comment by nonchip** [comment]
@strycore so `wontfix`? 0.o

**Comment by strycore** [comment]
@nonchip no, it's been closed by this commit which fixes the issue and will be releases with 0.4.19

**Comment by nonchip** [comment]
@strycore the commit does introduce a whole bunch of other issues, as already described above.
